### DISCUSSION TITLE
Small feature request for formatting widget.net

### DISCUSSION THREAD

**Discussion by Crystal4276** [qtile/qtile#3999] [Ideas] [answered]
I'd like to be able to change the formatting of speed transfer.
For instance instead of 2.56M, i would like to display 2.6M.
If I understood well the source code, there is no possibility to adjust this at the moment  
```python
        down = "{val:{max_len}.2f}".format(val=down, max_len=max_len_down)
        up = "{val:{max_len}.2f}".format(val=up, max_len=max_len_up)
        total = "{val:{max_len}.2f}".format(val=total, max_len=max_len_total)
```
Other widget like widget.CPU, or widget.Memory have this formatting option.
Could we add it to widget.Net ?

**Comment by elParaguayo** [comment, marked as answer]
PR #3487 should address this. It's been open for a long time and, to be honest, I completely forgot about it!

**Comment by Crystal4276** [comment]
If you need me for testing stuff, let me know !

**Reply by elParaguayo** [reply]
If you're happy to test it, that would  be great.

**Comment by Crystal4276** [comment]
If i have a problem, i don't know where to copy those two files:
https://github.com/qtile/qtile/blob/a43c1aed77e6909f84256fdc6f6c4f784cda7b74/test/widgets/docs_screenshots/ss_net.py
https://github.com/qtile/qtile/blob/a43c1aed77e6909f84256fdc6f6c4f784cda7b74/test/widgets/test_net.py

**Reply by elParaguayo** [reply]
You don't need to copy those. They're just for running tests and making our docs.

**Comment by Crystal4276** [comment]
Ok then the formatting works
`    format=" {up:2.1f}{up_suffix:<2} {down:2.1f}{down_suffix:<2}",`
![image](https://user-images.githubusercontent.com/27446196/202155830-f795caf3-fb35-4ad1-8e09-a841222bd1a6.png)

Though, I have trouble to understand what the parameter from `{up_suffix:<2} ` does ?

**Reply by elParaguayo** [reply]
`<2` means that the suffix should take up two character spaces and be left aligned. This is useful where users are using a monospace font and want a constant width for the widget.

**Comment by Crystal4276** [comment]
Hum wait, I was expecting with `2.1f` to provide 2 numbers before the period and 1 number after the period.
Instead i only get 1 number before the period.

To get two numbers before the period i need to use {up:4.1f}, is that normal ?
` format=" {up:4.1f}{up_suffix:<2} {down:4.1f}{down_suffix:<2}",`

![image](https://user-images.githubusercontent.com/27446196/202157048-37ee43b0-4263-4c96-82e6-3a1afe6ba355.png)

**Reply by elParaguayo** [reply]
`2.1f` means that the **total** string created by the field should be at least 2 characters long and there should be 1 number after the decimal. The 2 does not refer to the length before the decimal point.

**Reply by Crystal4276** [reply]
> `2.1f` means that the **total** string created by the field should be at least 2 characters long and there should be 1 number after the decimal. The 2 does not refer to the length before the decimal point.

Ok then there is an issue, because 3.1f and 2.1f provide the same number of characters.
4.1f and 5.1f are also the same.

**Reply by elParaguayo** [reply]
I've probably explained this badly.
```
>>>a = 1.234
>>>print(f"{a:2.1f}")
"1.2"
>>>print(f"{a:4.1f}")
" 1.2"
```
Note the space before the 1 in the second example. No additional digits are added because the `.1` says there can only be one number after the decimal place.

**Reply by Crystal4276** [reply]
> Note the space before the 1 in the second example. No additional digits are added because the `.1` says there can only be one number after the decimal place.

That's the part i don't understand. 
I was expecting **two** additional space with `4.1f` and only one additional space with `3.1f`.

Ok don't bother no worries, i'll probably need to further read on that.
Then i guess the widget.net is fine !

**Reply by elParaguayo** [reply]
If you're not using a monospace font, the spaces won't take up much space (and will take up less space than the numbers). What makes you think you're not getting the spaces?

**Reply by Crystal4276** [reply]
> If you're not using a monospace font, the spaces won't take up much space (and will take up less space than the numbers). What makes you think you're not getting the spaces?

Well i can show you:
` format=" {up:2.1f}{up_suffix:<2} {down:2.1f}{down_suffix:<2}",`
![image](https://user-images.githubusercontent.com/27446196/202187353-f34a31f5-d8af-4283-a5d0-4708891b0c8d.png)
` format=" {up:3.1f}{up_suffix:<2} {down:3.1f}{down_suffix:<2}",`
![image](https://user-images.githubusercontent.com/27446196/202187515-77a7a646-e481-4737-83cb-a27c4ca2d581.png)
`format=" {up:4.1f}{up_suffix:<2} {down:4.1f}{down_suffix:<2}",`
![image](https://user-images.githubusercontent.com/27446196/202187871-df54228b-dedc-4594-8a11-770249385e6e.png)

**Reply by elParaguayo** [reply]
That's the correct behaviour.

For the first one, the total string must be at least 2 characters, and it is: `1`, `.` and `7`.
For the second, it needs to be at least 3 characters, and it is: `3`, `.` and `1`.
Finally, for the fourth, it needs to be at least 4 characters. `3`, `.` and `2` are only 3 characters so you get the extra space at the start.

**Reply by Crystal4276** [reply]
Oh now i get it! 
I was missing the `.` 
Sorry I'm super slow.
Thanks for taking the time!

**Reply by elParaguayo** [reply]
You're welcome.

**Reply by Crystal4276** [reply]
If i want to keep one decimal as along as up is < 10MB and if ever up >= 10MB then i want to remove the decimal.
Is it some `if` statements to write or is there a clever way of formatting this directly ?

**Reply by elParaguayo** [reply]
That's not going to be possible without modifying the widget but that wouldn't be too hard.
